tssc/step_implementers/sign_container_image/curl_push.py

```diff
@@ -0,0 +1,208 @@
+        )
+        signature_server_password = self.get_config_value(
+            'container-image-signature-server-password'
+        )
+
+        # extract step results
+        previous_sign_container_image_step_results = self.get_step_results(
+            DefaultSteps.SIGN_CONTAINER_IMAGE
+        )
+        CurlPush.__verify_previous_sign_container_image_step_results(
+            previous_sign_container_image_step_results,
+            ['container-image-signature-file-path', 'container-image-signature-name']
+        )
+        container_image_signature_file_path, container_image_signature_name = itemgetter(
+            'container-image-signature-file-path',
+            'container-image-signature-name'
+        )(previous_sign_container_image_step_results)
+
+        # upload
+        container_image_signature_url, signature_file_md5, signature_file_sha1 = \
+            CurlPush.__curl_file(
+                container_image_signature_file_path=container_image_signature_file_path,
+                container_image_signature_name=container_image_signature_name,
+                signature_server_url=signature_server_url,
+                signature_server_username=signature_server_username,
+                signature_server_password=signature_server_password
+            )
+
+        return {
+            'container-image-signature-url': container_image_signature_url,
+            'container-image-signature-file-md5': signature_file_md5,
+            'container-image-signature-file-sha1': signature_file_sha1
+        }
+
+    @staticmethod
+    def __verify_previous_sign_container_image_step_results(step_results, keys):
+        """Verifies that the given step results has the expected keys"""
+        if step_results is None:
+            raise RuntimeError(f"Missing step results from {DefaultSteps.SIGN_CONTAINER_IMAGE}")
+
+        for key in keys:
+            result = step_results.get(key)
+            if result is None:
+                raise RuntimeError(
+                    f"Missing {key} step results from {DefaultSteps.SIGN_CONTAINER_IMAGE}"
+                )
+
+    @staticmethod
+    def __curl_file(
+            container_image_signature_file_path,
+            container_image_signature_name,
+            signature_server_url,
+            signature_server_username,
+            signature_server_password
+    ):
+        """Sends the signature file"""
+        # remove any trailing / from url
+        signature_server_url = re.sub(r'/$', '', signature_server_url)
+        container_image_signature_url = f"{signature_server_url}/{container_image_signature_name}"
+
+        # calculate hashes
+        with open(container_image_signature_file_path, 'rb') as container_image_signature_file:
+            container_image_signature_file_contents = container_image_signature_file.read()
+            signature_file_md5 = hashlib.md5(container_image_signature_file_contents).hexdigest()
```

Comment:
oh man I didn't expect to see this. :) Using md5 instead of hashlib.md5 is one of the more common crypto errors in python.

Reply:
haha I won't take credit for this, this was @itewk